<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header" style="float: left">
        <a class="navbar-brand" href="0_Forside.ipynb" target="_self"> <h2> &uarr; Tilbake til forsiden</h2></a>
    </div>
  </div>
</nav>

# Flyttall og nøyaktiget i beregninger

**Læringsmål:**

* Flyttallrepresentasjon
* Avrundingsfeil

## Intro til avrundignsfeil

Programmering kan være et svært nyttig verktøy til å utføre tunge matematiske beregnigner men det er viktig å være klar over at det også finnes begrensninger, og at disse kan gi resultat som kanskje ikke ser helt ut som man hadde forventet.

La oss ta for oss et eksempel:
>Basert på det vi har lært i geometrien vet vi at $\sin(180^\circ)=0$, eventuelt at $\sin(\pi)=0$ når vi måler vinkel i radianer. Når vi prøver å regne ut dette med bruk av Python får vi derimot et litt annet svar, slik kodecellen nedenfor demonstrerer:

In [1]:
from numpy import sin, pi

print(f"{sin(pi) = }")

sin(pi) = np.float64(1.2246467991473532e-16)


>Svaret vi får med Python kan ved første øyekast se veldig galt ut, men det er viktig å legge merke til at dette tallet bruker vitenskapelig notasjon. Dersom vi skulle skrevet ut dette som et ordinært desimaltall ville vi fått $15$ nuller bak desimaltegnet. Med andre ord er dette et veldig lite tall som er ***nesten*** er lik $0$, men måten datamaskiner utfører beregninger på har medført en aldri så liten regnefeil. Årsaken til dette er at en datamaskin ikke faktisk arbeider med den nøyaktige verdien til tallet $\pi$, men må rett og slett gjøre en avrunding da $\pi$ er et uendelig langt tall.

In [5]:
print(f"{pi = }")

pi = 3.141592653589793


## Representasjon av desimaltall
Python representerer altså tallet $\pi$ med 15 desimalers nøyaktighet, men i forrige oppgave gjorde vi beregninger med planck's konstant $h = 6.626\cdot10^{-34} \text{ J/s}$. Hvordan kan det ha seg at vi kan gjøre beregninger på et så lite tall, men samtidig rundes $\pi$ av til $15$ desimaler?

For å svare på dette spørsmålet skal vi ta for oss et tankeeksperiment; To fanger sitter i hver sin fengselcelle, mellom cellene er et vindu med fire små glassruter. Fangene kan kommunisere ved å skrive kun ett tallsymbol $0\text{-}9$ på hver av glassrutene. Gitt disse begrensningene kan fangene bruke 10-tallsystemet til å skrive ned alle heltall mellom $0$ og $9999$, men målet til fangene er å finne et sett med regler for hvordan de fire tallysmbolene tolkes som lar de formidle både små tall (f.eks. $0.001$) og store tall (f.eks. $100\ 000$) til hverandre.

Løsningen er å ta i bruk *vitenskapelig notasjon*. Dersom fangene blir enig om at at én av de fire glassrutene er eksponenten og de tre resterende rutene er signifikanden, betyr det at tallverdiene vi kan skrive ned kan bli så store som $999\cdot 10^9 = 999\ 000\ 000\ 000$. Men, legg merke til at når eksponenten blir så stor, så blir også avrundingsfeilen veldig stor fordi vi ender opp med å runde av til nærmeste milliard. 

For å klare å skrive ned tall som er mindre enn 1, må fangene ha muligheten til å skrive eksponenten med negativt fortegn. Én løsning er å trekke fra f.esk, $5$ fra tallsymbolet skrevet i eksponent-ruten, før man setter det inn i eksponentialuttrykket. Da vil denne ruten kunne tolkes som tallene $\{-5, \cdots 4\}$. Figuren nedenfor viser da hvordan symbolene  $\fbox{3}\fbox{1}\fbox{4}\fbox{3}$ vil kunne tolkes til å bety tallet $3.14$.

<img src="Figurer/Prisoners_float.png" style="width: 500px; margin-left: 10%" />

Til slutt kan vi generalisere det hele i en matteformel. Gitt fire variabler $\text{A}$, $\text{B}$, $\text{C}$ og $\text{D}$ som representerer de fire glassrutene, kan den ferdige oppskriften på konvertering fra 4 énsifrede tallsymbol ($0\text{-}9$) til en fleksibel tallverdi dermed oppsummeres som i formel $1$ nedenfor:

>$$\text{tall} = \left(\text{A} \cdot 100 + \text{B}\cdot 10 + \text{C}\right)\cdot 10 ^{\text{D-5}} \tag{1}$$
>der symbolene $\text{A}$, $\text{B}$, $\text{C}$ og $\text{D}$ er énsifrede heltall $[0\text{-}9]$.

### a)
Gitt regelen for tolking av tallverdi gitt i formel $1$, hva er den aller største tallverdien det går an å representere? (Med andre ord, hvilken tallverdi representerer symbolene $\fbox{9}\fbox{9}\fbox{9}\fbox{9}$ dersom de tolkes med metoden i formel $1$?) <br>Skriv svaret til variabelen `maks_tallverdi` i kodecellen under.

In [6]:
maks_tallverdi = (9 * 100 + 9 * 10 + 9) * 10 ** (9 - 5)
print(maks_tallverdi)

9990000


### b)
Gitt regelen for tolking av tallverdi gitt i formel $1$, hva er den aller minste tallverdien det går an å representere (annet enn $0.0$)?(Med andre ord, hvilken tallverdi representerer symbolene $\fbox{0}\fbox{0}\fbox{1}\fbox{0}$ dersom de tolkes med metoden i formel $1$?)  <br>Skriv svaret til variabelen `min_tallverdi` i kodecellen under.

In [8]:
min_tallverdi = (0 * 100 + 0 * 10 + 1) * 10 ** (0 - 5)
print(min_tallverdi)

1e-05


Generelt kan man si at metoden for å for å representere tall gitt i formel $1$ vil involvere å alltid runde av til 3 signifikante siffer, der det fjerde tallsymbolet bestemmer plasseringen av desimaltegnet. Slik vil eksempelvis $59\ 293$ avrundes til $59\ 300$ og representeres med symbolene $\fbox{5}\fbox{9}\fbox{3}\fbox{7}$, og $\pi$ avrundes til $3.14$ og representeres med symbolene $\fbox{3}\fbox{1}\fbox{4}\fbox{3}$. 

## Absolutt feil og relativ feil
I numerikk er absolutt feil et begrep vi bruker om avviket mellom en tilnærmet tallverdi og en eksakt tallverdi. Absolutt feil defineres som i formel $2$ nedenfor:

$$\text{absolutt feil} = \big| \text{avrundet tallverdi} - \text{eksakt tallverdi} \big| \tag{2}$$

Et annet mål på hvor nøyaktig en tallverdi er, er *relativ feil*, som er en indikasjon på hvor stor en avrundingsfeil er relativ til den eksakte verdien. Relativ feil kan regnes ut med formelen $3$ nedenfor:

$$\text{relativ feil} = \left|\frac{ \text{avrundet tallverdi} - \text{eksakt tallverdi} }{\text{eksakt tallverdi}}\right| \tag{3}$$

### c) 
Finn den avrundede verdien, og hva den den absolutte og relative feilen blir når man representerer en tilnærming til tallet $16\ 384$ ved hjelp av regneregelen i formel $1$. Du kan regne det ut enten for hånd eller med bruk av pythonkode.

In [9]:
tall = 16_384  # Nøyaktig tallverdi som skal representeres
avrundet_tallverdi = 16_400  # Vi beholder 3 signifikante siffer (runder av til nærmeste 100)

absolutt_feil = abs(tall - avrundet_tallverdi)
relativ_feil = abs((tall - avrundet_tallverdi) / tall)

print(f"{tall} rundes av til {avrundet_tallverdi}.")
print(f"Absolutt feil: {absolutt_feil}")
print(f"Relativ feil: {relativ_feil}")

16384 rundes av til 16400.
Absolutt feil: 16
Relativ feil: 0.0009765625


### d) 
Finn den avrundede verdien, og hva den den absolutte og relative feilen blir når man representerer en tilnærming til tallet $0.001436$ ved hjelp av metoden i formel $1$. Du kan regne det ut enten for hånd eller med bruk av pythonkode.

In [10]:
tall = 0.001438  # Nøyaktig tallverdi som skal representeres
avrundet_tallverdi = 0.00144  # Vi beholder 3 signifikante siffer (runder av til 5 desimals presisjon da de første to desimalene er 0)

absolutt_feil = abs(tall - avrundet_tallverdi)
relativ_feil = abs((tall - avrundet_tallverdi) / tall)

print(f"Avrundet tallverdi: {avrundet_tallverdi}")
print(f"Absolutt feil: {absolutt_feil}")
print(f"Relativ feil: {relativ_feil}")

Avrundet tallverdi: 0.00144
Absolutt feil: 2.0000000000000486e-06
Relativ feil: 0.001390820584144679


Måten vi har representert tallverdien i denne oppgaven er en forenkling av hvordan en datamaskin representerer desimaltall. Ettersom talleverdien representeres med et justerbart, eller "flytende" desimaltegn, kalles desimaltall typisk for [*flyttall*](https://docs.python.org/3/tutorial/floatingpoint.html). Som vi har sett i oppgave **c)** og **d)**, vil den absolutte feilen for et flyttall avhenge av størrelseordenen til den eksakte tallverdien, men den relative feilen vil være i samme størrelsesorden for både store og små tall. Dette er grunnen til at den absolutte avrundingsfeilen for $\pi$ er så mye større enn Planck's konstant $h = 6.626\cdot10^{-34} \text{ J/s}$.

Én detalj som gjør flyttall i Python mer kompliserte enn tankeeksperimentet vårt er at det er mange flere siffer i tallene. I tillegg bruker ikke datamaskiner titall-systemet, men det *binære* tallsystemet til å representere tallverdier. I neste oppgave skal vi se litt nærmere på hvordan binære tall fungerer, og hvordan dette påvirker oppførselen til flyttall i Python.

<br>
<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header" style="float: left">
      <a class="navbar-brand" href="6_Vitenskapelig_notasjon.ipynb" target="_self">&lt; Forrige side: <i>Vitenskapelig notasjon</i></a>
      </div>
    <div class="navbar-header" style="float: right">
      <a class="navbar-brand" href="9_Binaertall.ipynb" target="_self">Neste side: <i>Binære tall</i> &gt;</a>
    </div>
  </div>
</nav>